File: 2015/19/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const REGEX = /^([a-zA-Z]+) => ([a-zA-Z]+)$/;

// --- Functions ---
function shuffle(a) {
	for (let index = a.length - 1; index > 0; index--) {
		const j = Math.floor(Math.random() * (index + 1));
		const x = a[index];
		a[index] = a[j];
		a[j] = x;
	}
	return a;
}

// --- Variables ---
let map = {};
let pairs = [];
let molecule = "";
let first = true;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	if (line.length === 0) {
		first = false;
	} else {
		if (first) {
			let cap = line.match(REGEX);
			if (!map.hasOwnProperty(cap[1])) map[cap[1]] = [];
			map[cap[1]].push(cap[2]);
			pairs.push(cap[1] + "-" + cap[2]);
		} else {
			molecule = line;
		}
	}
});
await new Promise(res => reader.once("close", res));

// --- Puzzle 1 ---
const regex = new RegExp("(" + Object.keys(map).join("|") + ")", "g");
let set = new Set();
let cap;
while ((cap = regex.exec(molecule)) !== null) {
	const i = cap.index;
	const j = i + cap[1].length;
	for (const x of map[cap[1]]) {
		const s = molecule.substring(0, i) + x + molecule.substring(j);
		set.add(s);
	}
}
console.log("1. Distinct molecules can be created: ", set.size.toLocaleString());

// --- Puzzle 2 ---
let target = molecule.slice();
let steps = 0;
while (target !== "e") {
	let change = false;
	for (const x of pairs) {
		const t = x.split("-");
		const from = t[0];
		const to = t[1];
		if (target.includes(to)) {
			target = target.replace(to, from);
			change = true;
			break;
		}
	}

	// --- If not change was possible shuffle pairs and try again ---
	if (!change) {
		shuffle(pairs);
		target = molecule.slice();
		steps = 0;
		continue;
	}

	steps++;
}
console.log("2. Fewest number of steps: ", steps.toLocaleString());

File: 2015/19/python/main.py
import random
import re

# --- Constants --
INPUT_FILE = "../input/input.txt"
REGEX = r"^([a-zA-Z]+) => ([a-zA-Z]+)$"

# --- Variables ---
_map = {}
pairs = []
molecule = ""
first = True

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    if len(line) == 0:
        first = False
    else:
        if first:
            cap = re.search(REGEX, line)
            if cap[1] not in _map: _map[cap[1]] = []
            _map[cap[1]].append(cap[2])
            pairs.append((cap[1], cap[2]))
        else:
            molecule = line
file.close()

# --- Puzzle 1 ---
regex = "(" + "|".join(_map.keys()) + ")"
_set = set()
pattern = re.compile(regex)
for match in pattern.finditer(molecule):
    i = match.start()
    j = match.end()
    v = match[0]
    for x in _map[v]:
        s = molecule[0:i] + x + molecule[j:]
        _set.add(molecule[0:i] + x + molecule[j:])
print(f"1. Distinct molecules can be created: {len(_set):,d}")

# --- Puzzle 2 ---
target = molecule[:]
steps = 0
while target != "e":
    change = False
    for p in pairs:
        if p[1] in target:
            target = target.replace(p[1], p[0], 1)
            change = True
            break

    if not change:
        random.shuffle(pairs)
        target = molecule[:]
        steps = 0
        continue;

    steps += 1
print(f"2. Fewest number of steps: {steps:,d}")
